akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala

```diff
@@ -249,38 +270,49 @@ object ShardCoordinator {
    * 10 shards per node. One node may have 19 shards and others 10 without a rebalance occurring.
    *
    * The number of ongoing rebalancing processes can be limited by `maxSimultaneousRebalance`.
+   *
+   * During a rolling upgrade (when nodes with multiple application versions are present) allocating to
+   * old nodes are avoided.
+   *
+   * Not intended for user extension.
    */
   @SerialVersionUID(1L)
+  @DoNotInherit
   class LeastShardAllocationStrategy(rebalanceThreshold: Int, maxSimultaneousRebalance: Int)
-      extends ShardAllocationStrategy
+      extends AbstractLeastShardAllocationStrategy
       with Serializable {
 
     override def allocateShard(
         requester: ActorRef,
         shardId: ShardId,
         currentShardAllocations: Map[ActorRef, immutable.IndexedSeq[ShardId]]): Future[ActorRef] = {
-      val (regionWithLeastShards, _) = currentShardAllocations.minBy { case (_, v) => v.size }
-      Future.successful(regionWithLeastShards)
+      val decorated = regionEntriesFor(currentShardAllocations)
```

Comment:
Hmm, what scenario is it that region entries would be empty, when region is known, but the member is not, can that really happen? I mean, registration requires a network roundtrip, wouldn't that always take longer time than the change being visible from the cluster state view? (Should we hot-spin on it until it is seen, in that case?)

I'll rename variable, missed this comment in the previous review pass.

Reply:
I was thinking that it could happen since `regionEntriesFor` does the flatMap and has the comment
// if the member is unknown (very unlikely but not impossible) because of view not updated yet

However, I think you are right that it should be unlikely. Maybe more likely for the downing removal scenario than the joining. Would feel better to have the isEmpy check. I think the new strategy has that already.

Reply:
No empty check in the new strategy either, what do we do in that scenario, hot spin in hopes that the cluster view is updated? Schedule a delayed future retry?

Reply:
Added a delayed retry.